addon/components/course-materials.js

```diff
@@ -1,126 +1,156 @@
-    return new RSVPPromise(resolve =>{
-      const course = this.get('course');
-      course.get('learningMaterials').then(clms => {
-        map(clms.toArray(), courseLearningMaterial => {
-          return new RSVPPromise(resolve =>{
-            courseLearningMaterial.get('learningMaterial').then(learningMaterial => {
-              let obj = EmberObject.create({
-                title: learningMaterial.get('title'),
-                description: learningMaterial.get('description'),
-                author: learningMaterial.get('originalAuthor'),
-                type: learningMaterial.get('type'),
-                url: learningMaterial.get('url'),
-                citation: learningMaterial.get('citation'),
-              });
-              resolve(obj);
-            });
-          });
-        }).then(lmObjects => {
-          resolve(lmObjects);
-        });
-      });
+
+  actions: {
+    clmSortBy(prop) {
+      if (this.courseSort === prop) {
+        prop += ':desc';
+      }
+      this.onClmSort(prop);
+    },
+
+    slmSortBy(prop) {
+      if (this.sessionSort === prop) {
+        prop += ':desc';
+      }
+      this.onSlmSort(prop);
+    }
+  },
+
+  setClmQuery: task(function* (q) {
+    yield timeout(DEBOUNCE_DELAY);
+    this.set('clmQuery', q);
+  }).restartable(),
+
+  setSlmQuery: task(function* (q) {
+    yield timeout(DEBOUNCE_DELAY);
+    this.set('slmQuery', q);
+  }).restartable(),
+
+  async fetchCourseLearningMaterials() {
+    const clms = await this.course.get('learningMaterials');
+    const promises = clms.map((clm) => this.buildClmObject(clm));
+    return await all(promises);
+  },
+
+  async buildClmObject(clm) {
+    const lm = await clm.get('learningMaterial');
+    return EmberObject.create({
+      author: lm.originalAuthor,
+      citation: lm.citation,
+      description: lm.description,
+      title: lm.title,
+      type: lm.type,
+      url: lm.url
     });
-  }),
-  init(){
-    this._super(...arguments);
-    this.set('typesWithUrl', ['file', 'link']);
   },
+
+  async fetchSessionLearningMaterials() {
+    const sessions = await this.course.sessions;
+    const lms = await all(sessions.mapBy('learningMaterials'));
+    const slms = lms.reduce((flattened, obj) => {
+      return flattened.pushObjects(obj.toArray());
+    }, []);
+    const promises = slms.map((slm) => this.buildSlmObject(slm));
+    return await all(promises);
+  },
+
+  async buildSlmObject(slm) {
+    const lm = await slm.get('learningMaterial');
+    const session = await slm.session;
+    const firstOfferingDate = await session.firstOfferingDate;
+    return EmberObject.create({
+      author: lm.originalAuthor,
+      citation: lm.citation,
+      description: lm.description,
+      firstOfferingDate,
+      sessionTitle: session.title,
+      title: lm.title,
+      type: lm.type,
+      url: lm.url
+    });
+  },
+
+  filterClmo(clmo) {
+    const exp = new RegExp(this.clmQuery, 'gi');
```

Comment:
This query will need to be escaped before it can be used here. There is a cleanQuery util for this.  I also seem to recall that @stopfstedt was using a non-regex pattern for doing this kind of filtering and can probably point you in that direction instead.

Reply:
I just used what I extracted from below without changing anything.

https://github.com/ilios/common/blob/master/addon/components/course-materials.js#L67

I can look into it further if you feel that more is needed.

Reply:
Yes, please change. Unescaped regex has caused us problems in the past if users enter a `/`. Pre-existing bug and not your fault, but worth cleaning up here.